File: typescript/src/backtest.ts
/**
 * From one honest snapshot to a whole honest backtest.
 *
 * **The panel** — {@link buildPanel}
 * One snapshot per rebalance date, stacked. `coverage` reports how many series each
 * instant resolved — a falling count is a feed going dark, and it is invisible once the
 * rows are concatenated.
 *
 * **What the data said, and what it says now** — {@link restatementHistory}
 * The revision trail for one observation. A feature whose first print is routinely
 * revised is not the feature you backtested.
 *
 * **Would this backtest reproduce?** — {@link vintageDiff}
 * The same query at two knowledge times. Any series whose value changed was restated in
 * between — which means the run you did in March cannot be reproduced in June, and
 * neither result is wrong.
 *
 * **How late is the data, really?** — {@link publicationLag}
 * The median is the delay you should be modelling; the maximum is the one that breaks
 * you.
 *
 * **How much lookahead did you avoid?** — {@link leakageReport}
 * A large `withheld_share` is not a problem — it is the guard working.
 */

import {
  type LeakedRecord,
  type PitRecord,
  asOfSnapshot,
  leakageAudit,
  parseTimestampMs,
  validateRecords,
} from "./core.ts";

const MS_PER_DAY = 86_400_000;

export interface CoveragePoint {
  knowledge_time: string;
  valid_time: string;
  series: number;
  entities: number;
}

export interface Panel {
  rows: Array<PitRecord & { knowledge_time: string }>;
  coverage: CoveragePoint[];
  knowledge_times: string[];
  total_rows: number;
  min_series: number;
  max_series: number;
  complete: boolean;
}

export interface RevisionStep {
  record_id: string;
  revision: number;
  available_at: string;
  value: number;
  change_from_previous: number | null;
  lag_days: number;
}

export interface RestatementHistory {
  entity: string;
  feature: string;
  observation_time: string;
  versions: RevisionStep[];
  revision_count: number;
  first_value: number | null;
  final_value: number | null;
  total_drift: number | null;
  was_restated: boolean;
}

export interface VintageDiff {
  earlier_knowledge_time: string;
  later_knowledge_time: string;
  valid_time: string;
  series_compared: number;
  changed: Array<{
    entity: string;
    feature: string;
    observation_time: string;
    earlier_record_id: string;
    later_record_id: string;
    earlier_value: number;
    later_value: number;
    delta: number;
  }>;
  appeared: Array<{ entity: string; feature: string; value: number }>;
  reproducible: boolean;
  max_abs_delta: number | null;
}

export interface LagStats {
  count: number;
  min_days: number | null;
  median_days: number | null;
  mean_days: number | null;
  max_days: number | null;
}

export interface PublicationLag {
  overall: LagStats;
  first_print: LagStats;
  by_feature: Record<string, LagStats>;
  negative_lags: number;
}

export interface LeakageReport {
  knowledge_time: string;
  valid_time: string;
  total_records: number;
  valid_time_matches: number;
  withheld: number;
  withheld_share: number;
  by_entity: Record<string, number>;
  by_feature: Record<string, number>;
  clean: boolean;
}

const compare = (a: string, b: string) => (a < b ? -1 : a > b ? 1 : 0);

const median = (values: number[]): number => {
  const sorted = [...values].sort((a, b) => a - b);
  const middle = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0
    ? (sorted[middle - 1]! + sorted[middle]!) / 2
    : sorted[middle]!;
};

const sortedRecord = <T>(record: Record<string, T>): Record<string, T> =>
  Object.fromEntries(Object.entries(record).sort(([a], [b]) => compare(a, b)));

/** One as-of snapshot per knowledge time, stacked into a point-in-time panel. */
export function buildPanel(
  records: Iterable<unknown>,
  knowledgeTimes: readonly string[],
  validTimes?: readonly (string | null)[] | null,
): Panel {
  const recordList = validateRecords(records);
  const cutoffs = [...knowledgeTimes];
  if (cutoffs.length === 0) throw new Error("knowledge_times must not be empty");
  const validList = validTimes ? [...validTimes] : cutoffs.map(() => null);
  if (validList.length !== cutoffs.length) {
    throw new Error("valid_times must match knowledge_times in length");
  }

  const rows: Array<PitRecord & { knowledge_time: string }> = [];
  const coverage: CoveragePoint[] = [];

  cutoffs.forEach((knowledgeTime, index) => {
    const validTime = validList[index] ?? null;
    const snapshot = asOfSnapshot(recordList, knowledgeTime, validTime);
    for (const record of snapshot) rows.push({ ...record, knowledge_time: knowledgeTime });
    coverage.push({
      knowledge_time: knowledgeTime,
      valid_time: validTime || knowledgeTime,
      series: snapshot.length,
      entities: new Set(snapshot.map((row) => row.entity)).size,
    });
  });

  const seriesCounts = coverage.map((point) => point.series);
  return {
    rows,
    coverage,
    knowledge_times: cutoffs,
    total_rows: rows.length,
    // A falling series count is a feed going dark, and it disappears the moment the
    // rows are concatenated into one frame.
    min_series: Math.min(...seriesCounts),
    max_series: Math.max(...seriesCounts),
    complete: new Set(seriesCounts).size === 1,
  };
}

/**
 * The full revision trail for one observation: every version and when it landed.
 *
 * `total_drift` is the distance from the first print to the last — the number that says
 * whether the feature you backtested is the feature you have.
 */
export function restatementHistory(
  records: Iterable<unknown>,
  entity: string,
  feature: string,
  observationTime: string,
): RestatementHistory {
  const rows = validateRecords(records);
  const target = parseTimestampMs(observationTime, "observation_time");

  const versions = rows
    .filter(
      (row) =>
        row.entity === entity &&
        row.feature === feature &&
        parseTimestampMs(row.observation_time) === target,
    )
    .sort((a, b) => {
      const arrival = parseTimestampMs(a.available_at) - parseTimestampMs(b.available_at);
      if (arrival !== 0) return arrival;
      if (a.revision !== b.revision) return a.revision - b.revision;
      return compare(a.record_id, b.record_id);
    });

  const trail: RevisionStep[] = versions.map((row, index) => {
    const previous = index ? versions[index - 1]! : null;
    return {
      record_id: row.record_id,
      revision: row.revision,
      available_at: row.available_at,
      value: row.value,
      change_from_previous: previous ? row.value - previous.value : null,
      lag_days: (parseTimestampMs(row.available_at) - target) / MS_PER_DAY,
    };
  });

  const first = versions.length ? versions[0]!.value : null;
  const last = versions.length ? versions.at(-1)!.value : null;
  return {
    entity,
    feature,
    observation_time: observationTime,
    versions: trail,
    revision_count: trail.length,
    first_value: first,
    final_value: last,
    // How far the number moved after you first acted on it.
    total_drift: versions.length ? last! - first! : null,
    was_restated: trail.length > 1,
  };
}

/**
 * Diff two vintages of the same query: would this backtest reproduce?
 *
 * Both snapshots use the **same** `validTime`, so the comparison isolates *what you
 * learned in between* rather than mixing in new observations.
 */
export function vintageDiff(
  records: Iterable<unknown>,
  earlierKnowledgeTime: string,
  laterKnowledgeTime: string,
  validTime?: string | null,
): VintageDiff {
  const recordList = validateRecords(records);
  const earlierMs = parseTimestampMs(earlierKnowledgeTime, "earlier_knowledge_time");
  const laterMs = parseTimestampMs(laterKnowledgeTime, "later_knowledge_time");
  if (laterMs < earlierMs) {
    throw new Error("later_knowledge_time must not precede earlier_knowledge_time");
  }

  const sharedValid = validTime || earlierKnowledgeTime;
  const keyOf = (row: PitRecord) => `${row.entity} ${row.feature}`;
  const earlier = new Map(
    asOfSnapshot(recordList, earlierKnowledgeTime, sharedValid).map((r) => [keyOf(r), r]),
  );
  const later = new Map(
    asOfSnapshot(recordList, laterKnowledgeTime, sharedValid).map((r) => [keyOf(r), r]),
  );

  const changed: VintageDiff["changed"] = [];
  const appeared: VintageDiff["appeared"] = [];
  const keys = [...new Set([...earlier.keys(), ...later.keys()])].sort(compare);

  for (const key of keys) {
    const before = earlier.get(key);
    const after = later.get(key);
    if (!before) {
      appeared.push({ entity: after!.entity, feature: after!.feature, value: after!.value });
      continue;
    }
    if (!after || before.record_id === after.record_id) continue;
    changed.push({
      entity: before.entity,
      feature: before.feature,
      observation_time: before.observation_time,
      earlier_record_id: before.record_id,
      later_record_id: after.record_id,
      earlier_value: before.value,
      later_value: after.value,
      delta: after.value - before.value,
    });
  }

  return {
    earlier_knowledge_time: earlierKnowledgeTime,
    later_knowledge_time: laterKnowledgeTime,
    valid_time: sharedValid,
    series_compared: keys.length,
    changed,
    // Series absent at the earlier vintage: data that had not arrived at all yet.
    appeared,
    // True when the same window produces the same numbers at both vintages.
    reproducible: changed.length === 0 && appeared.length === 0,
    max_abs_delta: changed.length
      ? Math.max(...changed.map((row) => Math.abs(row.delta)))
      : null,
  };
}

/**
 * The distribution of `available_at − observation_time`, overall and per feature.
 *
 * The median is the delay you should be modelling. The **maximum** is the one that
 * breaks you. Only first prints are counted in `first_print`, since a later revision's
 * lag measures the restatement, not the publication.
 */
export function publicationLag(records: Iterable<unknown>): PublicationLag {
  const rows = validateRecords(records);
  if (rows.length === 0) throw new Error("records must not be empty");

  const stats = (values: number[]): LagStats =>
    values.length === 0
      ? { count: 0, min_days: null, median_days: null, mean_days: null, max_days: null }
      : {
          count: values.length,
          min_days: Math.min(...values),
          median_days: median(values),
          mean_days: values.reduce((a, b) => a + b, 0) / values.length,
          max_days: Math.max(...values),
        };

  const lag = (row: PitRecord) =>
    (parseTimestampMs(row.available_at) - parseTimestampMs(row.observation_time)) / MS_PER_DAY;

  const allLags = rows.map(lag);
  const firstPrints = rows.filter((row) => row.revision === 0).map(lag);

  const byFeature: Record<string, number[]> = {};
  for (const row of rows) (byFeature[row.feature] ??= []).push(lag(row));

  return {
    overall: stats(allLags),
    // The lag that decides whether the feature is usable at all.
    first_print: stats(firstPrints),
    by_feature: sortedRecord(
      Object.fromEntries(Object.entries(byFeature).map(([k, v]) => [k, stats(v)])),
    ),
    negative_lags: allLags.filter((value) => value < 0).length,
  };
}

/**
 * Roll up the leakage audit: how much lookahead the guard actually withheld.
 *
 * A large `withheld_share` is not a problem — it is the guard working, and it is the
 * figure to put in front of anyone who thinks point-in-time filtering is a formality.
 */
export function leakageReport(
  records: Iterable<unknown>,
  knowledgeTime: string,
  validTime?: string | null,
): LeakageReport {
  const rows = validateRecords(records);
  const leaked: LeakedRecord[] = leakageAudit(rows, knowledgeTime, validTime);
  const validCutoff = parseTimestampMs(validTime || knowledgeTime, "valid_time");
  const validRecords = rows.filter(
    (row) => parseTimestampMs(row.observation_time) <= validCutoff,
  );

  const byEntity: Record<string, number> = {};
  const byFeature: Record<string, number> = {};
  for (const row of leaked) {
    byEntity[row.entity] = (byEntity[row.entity] ?? 0) + 1;
    byFeature[row.feature] = (byFeature[row.feature] ?? 0) + 1;
  }

  return {
    knowledge_time: knowledgeTime,
    valid_time: validTime || knowledgeTime,
    total_records: rows.length,
    // What a valid-time-only filter would have admitted.
    valid_time_matches: validRecords.length,
    withheld: leaked.length,
    withheld_share: validRecords.length ? leaked.length / validRecords.length : 0,
    by_entity: sortedRecord(byEntity),
    by_feature: sortedRecord(byFeature),
    clean: leaked.length === 0,
  };
}

File: python/src/fintech_point_in_time/backtest.py
"""From one honest snapshot to a whole honest backtest.

The core module answers "what did I know at this instant". A backtest needs that answer
at *every* instant, plus the evidence that the answer is stable. This module supplies
both, and each function exists because of a specific way point-in-time discipline is lost
after the guard is already in place.

**The panel** — :func:`build_panel`
One snapshot per rebalance date, stacked. This is the object a backtest actually
consumes, and building it by looping over ``as_of_snapshot`` is the part teams get right;
what they miss is that the panel should be **built once and reused**, because a panel
rebuilt after a restatement is a different panel.

**What the data said, and what it says now** — :func:`restatement_history`
The revision trail for one series and one observation: every version, when it arrived,
and how far it moved. A feature whose first print is routinely revised by 3% is not the
feature you backtested, whatever the final values say.

**Would this backtest reproduce?** — :func:`vintage_diff`
Run the same query at two different knowledge times and diff the panels. Any series whose
value changed was **restated between the two runs** — which means the backtest you ran in
March cannot be reproduced in June, and neither result is wrong.

**How late is the data, really?** — :func:`publication_lag`
The distribution of ``available_at − observation_time``. The median is the delay you
should be modelling; the maximum is the one that breaks you. A feature with a 2-day
median and a 40-day tail cannot be used on a 3-day rebalance no matter what the median
says.

**How much lookahead did you avoid?** — :func:`leakage_report`
Rolls up :func:`~fintech_point_in_time.core.leakage_audit` by entity and feature, and
reports the share of otherwise-valid records that were correctly withheld. On the
reference fixture that is a large number, which is the point: the guard is doing
something, and this is how you show it.
"""

from __future__ import annotations

from statistics import mean, median
from typing import Any, Iterable, Mapping, Sequence

from .core import (
    as_of_snapshot,
    leakage_audit,
    parse_timestamp_ms,
    validate_records,
)

__all__ = [
    "build_panel",
    "leakage_report",
    "publication_lag",
    "restatement_history",
    "vintage_diff",
]

_MS_PER_DAY = 86_400_000


def build_panel(
    records: Iterable[Mapping[str, Any]],
    knowledge_times: Sequence[str],
    valid_times: Sequence[str] | None = None,
) -> dict[str, Any]:
    """One as-of snapshot per knowledge time, stacked into a point-in-time panel.

    Args:
        records: The bitemporal record set.
        knowledge_times: The rebalance instants, in the order you want them.
        valid_times: Optional per-instant valid-time cutoffs. Must match
            ``knowledge_times`` in length when supplied.

    Returns:
        ``rows`` carries every snapshot record stamped with the ``knowledge_time`` that
        produced it, so the panel remains self-describing after it leaves this function.
        ``coverage`` reports how many series each instant resolved — a falling count is
        a feed going dark, and it is invisible once the rows are concatenated.
    """

    record_list = validate_records(records)
    cutoffs = list(knowledge_times)
    if not cutoffs:
        raise ValueError("knowledge_times must not be empty")
    valid_list = list(valid_times) if valid_times is not None else [None] * len(cutoffs)
    if len(valid_list) != len(cutoffs):
        raise ValueError("valid_times must match knowledge_times in length")

    rows: list[dict[str, Any]] = []
    coverage: list[dict[str, Any]] = []
    for knowledge_time, valid_time in zip(cutoffs, valid_list):
        snapshot = as_of_snapshot(record_list, knowledge_time, valid_time)
        for record in snapshot:
            rows.append({**record, "knowledge_time": knowledge_time})
        coverage.append(
            {
                "knowledge_time": knowledge_time,
                "valid_time": valid_time or knowledge_time,
                "series": len(snapshot),
                "entities": len({row["entity"] for row in snapshot}),
            }
        )

    series_counts = [point["series"] for point in coverage]
    return {
        "rows": rows,
        "coverage": coverage,
        "knowledge_times": cutoffs,
        "total_rows": len(rows),
        # A falling series count is a feed going dark, and it disappears the moment the
        # rows are concatenated into one frame.
        "min_series": min(series_counts),
        "max_series": max(series_counts),
        "complete": len(set(series_counts)) == 1,
    }


def restatement_history(
    records: Iterable[Mapping[str, Any]],
    entity: str,
    feature: str,
    observation_time: str,
) -> dict[str, Any]:
    """The full revision trail for one observation: every version and when it landed.

    ``total_drift`` is the distance from the first print to the last. A feature whose
    first print is routinely revised is not the feature you backtested, and this is the
    number that says so.
    """

    rows = validate_records(records)
    target = parse_timestamp_ms(observation_time, "observation_time")

    versions = sorted(
        (
            row
            for row in rows
            if row["entity"] == entity
            and row["feature"] == feature
            and parse_timestamp_ms(row["observation_time"]) == target
        ),
        key=lambda row: (parse_timestamp_ms(row["available_at"]), row["revision"], row["record_id"]),
    )

    trail: list[dict[str, Any]] = []
    for index, row in enumerate(versions):
        previous = versions[index - 1] if index else None
        trail.append(
            {
                "record_id": row["record_id"],
                "revision": row["revision"],
                "available_at": row["available_at"],
                "value": row["value"],
                "change_from_previous": (
                    row["value"] - previous["value"] if previous is not None else None
                ),
                "lag_days": (
                    parse_timestamp_ms(row["available_at"]) - target
                ) / _MS_PER_DAY,
            }
        )

    first = versions[0]["value"] if versions else None
    last = versions[-1]["value"] if versions else None
    return {
        "entity": entity,
        "feature": feature,
        "observation_time": observation_time,
        "versions": trail,
        "revision_count": len(trail),
        "first_value": first,
        "final_value": last,
        # How far the number moved after you first acted on it.
        "total_drift": (last - first) if versions else None,
        "was_restated": len(trail) > 1,
    }


def vintage_diff(
    records: Iterable[Mapping[str, Any]],
    earlier_knowledge_time: str,
    later_knowledge_time: str,
    valid_time: str | None = None,
) -> dict[str, Any]:
    """Diff two vintages of the same query: would this backtest reproduce?

    Both snapshots use the **same** ``valid_time`` — defaulting to the earlier knowledge
    time — so the comparison isolates *what you learned in between* rather than mixing in
    new observations. Any series whose value changed was restated between the two runs.

    That is the reproducibility answer: a backtest run in March and re-run in June over
    the identical window can legitimately disagree, and neither result is wrong.
    """

    record_list = validate_records(records)
    earlier_ms = parse_timestamp_ms(earlier_knowledge_time, "earlier_knowledge_time")
    later_ms = parse_timestamp_ms(later_knowledge_time, "later_knowledge_time")
    if later_ms < earlier_ms:
        raise ValueError("later_knowledge_time must not precede earlier_knowledge_time")

    shared_valid = valid_time or earlier_knowledge_time
    earlier = {
        (row["entity"], row["feature"]): row
        for row in as_of_snapshot(record_list, earlier_knowledge_time, shared_valid)
    }
    later = {
        (row["entity"], row["feature"]): row
        for row in as_of_snapshot(record_list, later_knowledge_time, shared_valid)
    }

    changed: list[dict[str, Any]] = []
    appeared: list[dict[str, Any]] = []
    for key in sorted(earlier.keys() | later.keys()):
        before = earlier.get(key)
        after = later.get(key)
        if before is None:
            appeared.append(
                {"entity": key[0], "feature": key[1], "value": after["value"]}
            )
            continue
        if after is None or before["record_id"] == after["record_id"]:
            continue
        changed.append(
            {
                "entity": key[0],
                "feature": key[1],
                "observation_time": before["observation_time"],
                "earlier_record_id": before["record_id"],
                "later_record_id": after["record_id"],
                "earlier_value": before["value"],
                "later_value": after["value"],
                "delta": after["value"] - before["value"],
            }
        )

    return {
        "earlier_knowledge_time": earlier_knowledge_time,
        "later_knowledge_time": later_knowledge_time,
        "valid_time": shared_valid,
        "series_compared": len(earlier.keys() | later.keys()),
        "changed": changed,
        # Series absent at the earlier vintage: data that had not arrived at all yet.
        "appeared": appeared,
        # True when the same window produces the same numbers at both vintages.
        "reproducible": not changed and not appeared,
        "max_abs_delta": (
            max(abs(float(row["delta"])) for row in changed) if changed else None
        ),
    }


def publication_lag(records: Iterable[Mapping[str, Any]]) -> dict[str, Any]:
    """The distribution of ``available_at − observation_time``, overall and per feature.

    The median is the delay you should be modelling. The **maximum** is the one that
    breaks you: a feature with a two-day median and a forty-day tail cannot be used on a
    three-day rebalance, whatever the median says.

    Only first prints (``revision == 0``) are counted by default in ``first_print``,
    since a later revision's lag measures the restatement, not the publication.
    """

    rows = validate_records(records)
    if not rows:
        raise ValueError("records must not be empty")

    def stats(values: list[float]) -> dict[str, Any]:
        if not values:
            return {"count": 0, "min_days": None, "median_days": None,
                    "mean_days": None, "max_days": None}
        return {
            "count": len(values),
            "min_days": min(values),
            "median_days": median(values),
            "mean_days": mean(values),
            "max_days": max(values),
        }

    def lag(row: Mapping[str, Any]) -> float:
        return (
            parse_timestamp_ms(row["available_at"])
            - parse_timestamp_ms(row["observation_time"])
        ) / _MS_PER_DAY

    all_lags = [lag(row) for row in rows]
    first_prints = [lag(row) for row in rows if row["revision"] == 0]

    by_feature: dict[str, list[float]] = {}
    for row in rows:
        by_feature.setdefault(str(row["feature"]), []).append(lag(row))

    return {
        "overall": stats(all_lags),
        # The lag that decides whether the feature is usable at all.
        "first_print": stats(first_prints),
        "by_feature": {
            feature: stats(values) for feature, values in sorted(by_feature.items())
        },
        "negative_lags": sum(1 for value in all_lags if value < 0),
    }


def leakage_report(
    records: Iterable[Mapping[str, Any]],
    knowledge_time: str,
    valid_time: str | None = None,
) -> dict[str, Any]:
    """Roll up the leakage audit: how much lookahead the guard actually withheld.

    ``withheld_share`` is over the records a valid-time-only filter would have admitted.
    A large number is not a problem — it is the guard working, and it is the figure to
    put in front of anyone who thinks point-in-time filtering is a formality.
    """

    rows = validate_records(records)
    leaked = leakage_audit(rows, knowledge_time, valid_time)
    valid_cutoff = parse_timestamp_ms(valid_time or knowledge_time, "valid_time")
    valid_records = [
        row for row in rows if parse_timestamp_ms(row["observation_time"]) <= valid_cutoff
    ]

    by_entity: dict[str, int] = {}
    by_feature: dict[str, int] = {}
    for row in leaked:
        by_entity[str(row["entity"])] = by_entity.get(str(row["entity"]), 0) + 1
        by_feature[str(row["feature"])] = by_feature.get(str(row["feature"]), 0) + 1

    return {
        "knowledge_time": knowledge_time,
        "valid_time": valid_time or knowledge_time,
        "total_records": len(rows),
        # What a valid-time-only filter would have admitted.
        "valid_time_matches": len(valid_records),
        "withheld": len(leaked),
        "withheld_share": (len(leaked) / len(valid_records)) if valid_records else 0.0,
        "by_entity": dict(sorted(by_entity.items())),
        "by_feature": dict(sorted(by_feature.items())),
        "clean": not leaked,
    }
